### PR TITLE
MAE Pretraining: Self-Supervised Geometry Encoder Initialization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver encoder currently learns two things simultaneously: (1) spatial reasoning about mesh structure and foil geometry, and (2) flow physics prediction. Pretraining the encoder using Masked Autoencoder (MAE) self-supervision on mesh geometry only — before supervised training — separates these objectives. For the OOD NACA6416 tandem case, a geometry-pretrained encoder may generalize better because it has learned a richer geometric representation covering ALL available geometries.

**GeoMAE (arXiv:2304.06911, CVPR 2023)** applied masked geometric reconstruction (surface normals, curvatures, centroids) to point cloud pretraining and showed significant improvements on downstream tasks. The Transolver's slice-attention mechanism is structurally similar to set-abstraction in point cloud networks.

**Key insight:** The model struggles on NACA6416 because it's geometrically OOD. If the encoder has already learned to represent diverse foil geometries through self-supervised pretraining, the supervised training can focus on learning flow physics rather than geometric structure. This decoupling should reduce the OOD gap.

### Step 1: MAE pretraining mode

Add a `--mae_pretrain` mode that runs before supervised training. The pretraining loop:

1. For each training batch, randomly mask 25% of nodes (replace their input features with a learned `[MASK]` token)
2. Run the Transolver encoder to get node embeddings for ALL nodes (including masked)
3. Run a lightweight decoder MLP (2 layers) to reconstruct the ORIGINAL input features at masked positions only
4. Loss: L1 reconstruction on masked node features

```python
class MAEDecoder(nn.Module):
    """Lightweight decoder for masked autoencoder pretraining."""
    def __init__(self, hidden_dim=192, output_dim=24):
        super().__init__()
        self.decoder = nn.Sequential(
            nn.Linear(hidden_dim, hidden_dim),
            nn.GELU(),
            nn.Linear(hidden_dim, output_dim)
        )

    def forward(self, h, mask):
        """
        h: (B, N, D) encoder hidden states
        mask: (B, N) boolean mask (True = masked nodes)
        Returns: reconstructed features at masked positions
        """
        masked_h = h[mask]  # (M, D)
        return self.decoder(masked_h)  # (M, output_dim)
```

**Masking strategy:**
```python
def random_mask(B, N, mask_ratio=0.25, device='cuda'):
    """Random node masking. Returns boolean mask."""
    mask = torch.rand(B, N, device=device) < mask_ratio
    return mask

# In pretraining forward pass:
mask = random_mask(B, N, mask_ratio=0.25)
# Replace masked node features with learned mask token
mask_token = nn.Parameter(torch.zeros(1, 1, input_dim))  # learned
x_masked = x.clone()
x_masked[mask] = mask_token.expand(mask.sum(), -1)
# Run encoder
h = encoder(x_masked)
# Reconstruct
x_recon = mae_decoder(h, mask)
# Loss: only on masked nodes
mae_loss = F.l1_loss(x_recon, x[mask][:, :output_dim])
```

### Step 2: Pretraining schedule

Run MAE pretraining for **20 epochs** before supervised training. This costs ~25 minutes (at ~70s/epoch), leaving ~155 minutes for supervised training (~130 epochs).

**IMPORTANT:** Use the SAME data loader as supervised training (including augmentation). The pretraining sees the same data distribution.

**Optimizer for pretraining:** Use the same Lion optimizer with lr=2e-4, but NO cosine annealing during pretraining (use constant LR). Reset the optimizer and scheduler for supervised training.

### Step 3: Supervised training with pretrained weights

After pretraining:
1. Save the encoder weights
2. Discard the MAE decoder (not needed for supervised training)
3. Initialize the full model with pretrained encoder weights
4. Run normal supervised training from epoch 0 with cosine annealing T_max=160
5. The pretrained weights are fine-tuned, not frozen

```python
if args.mae_pretrain:
    # Phase 1: MAE pretraining
    print("Phase 1: MAE pretraining...")
    mae_decoder = MAEDecoder(hidden_dim=192, output_dim=24).to(device)
    mae_optimizer = Lion(
        list(model.parameters()) + list(mae_decoder.parameters()),
        lr=2e-4, weight_decay=5e-5
    )
    for epoch in range(args.mae_pretrain_epochs):
        for batch in train_loader:
            mae_optimizer.zero_grad()
            mask = random_mask(B, N, 0.25)
            # ... forward pass with masking ...
            mae_loss.backward()
            mae_optimizer.step()

    # Discard MAE decoder and optimizer
    del mae_decoder, mae_optimizer

    # Phase 2: Normal supervised training
    print("Phase 2: Supervised training with pretrained encoder...")
    # Re-create optimizer and scheduler from scratch
    optimizer = Lion(model.parameters(), lr=2e-4, weight_decay=5e-5)
    scheduler = CosineAnnealingLR(optimizer, T_max=160)
    # ... proceed with normal training loop ...
```

### Step 4: Flags

- `--mae_pretrain`: Enable MAE pretraining phase before supervised training
- `--mae_pretrain_epochs 20`: Number of MAE pretraining epochs (default 20)
- `--mae_mask_ratio 0.25`: Fraction of nodes to mask (default 0.25)

### Step 5: Run experiments

```bash
# Config A: 20 pretrain epochs, seed 42
cd cfd_tandemfoil && python train.py --agent alphonse --seed 42 \
  --wandb_name "alphonse/mae-pretrain-20ep-s42" \
  --wandb_group "alphonse/mae-pretrain" \
  --mae_pretrain --mae_pretrain_epochs 20 --mae_mask_ratio 0.25 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias

# Config B: 10 pretrain epochs (faster, more supervised epochs)
# Same but --mae_pretrain_epochs 10
# Seed 73: change --seed 73
```

### Key implementation notes

1. **Mask token:** A single learned vector (24-dim) that replaces masked node features. Initialize to zero. This tells the encoder "this node's features are unknown — reconstruct from context."

2. **Reconstruction targets:** Reconstruct all 24 input channels at masked positions. The encoder must learn to infer geometric features (coordinates, DSDF, zone) from surrounding context — exactly the spatial reasoning needed for OOD generalization.

3. **torch.compile:** The masking operation (boolean indexing) may not be compile-friendly. If compile fails, run pretraining without compile and enable compile for supervised training only.

4. **VRAM:** The MAE decoder adds ~37K params (~150KB). The masking reduces the effective sequence length by 25%, so pretraining may actually use LESS VRAM than supervised training.

5. **Logging:** Log `mae_pretrain/loss` to W&B during pretraining. This helps verify the encoder is learning meaningful representations.

6. **Encoder includes SRF head:** The pretrained weights include the backbone BUT NOT the SRF head or prediction heads (those are task-specific). Only pretrain the TransolverBlock layers and the input projection.

7. **Epoch counting:** The pretraining epochs are separate from supervised epochs. Log them as `mae_epoch` in W&B. The supervised training starts at epoch 0 with a fresh optimizer and cosine schedule.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.05** | < 13.05 |
| p_oodc | **7.70** | < 7.70 |
| **p_tan** | **28.60** | **< 28.60** |
| p_re | **6.55** | < 6.55 |

**Baseline W&B runs:** d7l91p0x (s42, p_tan=28.9), j9btfx09 (s73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — MAE Pretraining

Implemented masked autoencoder pretraining: randomly mask 25% of mesh nodes' input features, replace with learned mask token, run through encoder, decode hidden→reconstruct standardized features with L1 loss. After pretraining, optimizer and scheduler are recreated from scratch for supervised training.

### Results

| Config | Seed | p_in | p_oodc | p_tan | p_re | MAE time | W&B |
|--------|------|------|--------|-------|------|----------|-----|
| 20ep pretrain | 42 | 20.57 | 16.21 | 39.46 | 10.63 | 33.1 min | `8a4c3ojp` |
| 20ep pretrain | 73 | 25.08 | 18.00 | 45.32 | 11.51 | 33.2 min | `lr8mow5f` |
| 10ep pretrain | 42 | 27.74 | 26.59 | 47.86 | 17.68 | 16.6 min | `twf3alql` |
| 10ep pretrain | 73 | 17.35 | 15.87 | 42.56 | 9.54 | 16.6 min | `bv9rangq` |
| **Baseline** | — | **13.05** | **7.70** | **28.60** | **6.55** | — | |

### MAE Pretrain Loss

Loss decreased from ~0.50 → 0.36 over 20 epochs, confirming the pretraining objective was working. The encoder learned to reconstruct masked features, but this did NOT help downstream.

### Analysis — **Strong negative result**

All metrics are **2-5x worse than baseline**. This is a catastrophic failure, not a minor regression.

**Why it failed:**

1. **Pretraining corrupts the encoder for supervised learning.** The MAE objective pushes weights toward feature reconstruction (spatial interpolation), which is fundamentally different from flow physics prediction. The supervised training starts from a bad initialization — worse than random — because the encoder is biased toward copying/interpolating geometry rather than learning physics.

2. **Reduced supervised training budget.** 10ep pretrain costs 16.6 min (leaving ~135 epochs). 20ep costs 33.1 min (leaving ~122 epochs). Both are fewer than the ~155 epochs the baseline gets, compounding the bad initialization problem.

3. **All output heads start from scratch with a corrupted encoder.** The prediction heads (Ux, Uy, p) and refinement heads are randomly initialized but receive input from an encoder biased toward reconstruction rather than physics.

4. **The dataset is too small for self-supervised pretraining to help.** With only 1322 training samples, the encoder can learn the geometry distribution quickly during supervised training — there's no advantage to pretraining on the same data.

### Conclusion

MAE pretraining is not suitable for this problem — the training set is small, the wall-clock budget is tight, and the reconstruction objective conflicts with the downstream physics objective. Dead end.

### Suggested alternatives

If geometry-aware initialization is still of interest:
1. **Coordinate-only noise denoising** (denoise jittered coordinates) as a much lighter auxiliary loss alongside supervised training — no separate phase needed
2. **Simply longer training with random init** — the baseline already converges well within 155 epochs

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2185 because Strong negative result — all configs are 2-5x worse than baseline. The student's analysis is correct:

1. **MAE pretraining corrupts the encoder** for downstream physics prediction — the reconstruction objective biases weights toward spatial interpolation rather than flow physics.
2. **Reduced training budget** (10-20 epochs lost to pretraining) compounds the bad initialization.
3. **Dataset too small** (1322 samples) for self-supervised pretraining to provide useful geometry priors — supervised training already learns geometry efficiently.

The pretraining loss decreasing (0.50→0.36) confirmed the implementation was correct, making this a clean conceptual failure rather than an implementation bug. Dead end. GPU time is better spent on fresh directions.

Thanks alphonse — the thorough analysis of why it failed is exactly what we need..
